compiler/passes/normalize.cpp

```diff
@@ -1293,14 +1301,163 @@ static bool isVoidReturn(CallExpr* call) {
+
+      return true;
+    }
+  }
+
+  return false;
+}
+
+static void insertDomainCheck(Expr* actualRet, CallExpr* retVar,
+                              Expr* domExpr) {
+  VarSymbol* unrefTemp = newTemp();
+  retVar->insertBefore(new DefExpr(unrefTemp));
+  CallExpr* makeArray = new CallExpr(PRIM_MOVE, unrefTemp,
+                                     new CallExpr("chpl__unref",
+                                                  actualRet->copy()));
+  retVar->insertBefore(makeArray);
+
+  VarSymbol* actualDom = newTemp();
+  retVar->insertBefore(new DefExpr(actualDom));
+  CallExpr* exprDom = new CallExpr("_dom", gMethodToken, unrefTemp);
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, actualDom, exprDom));
+
+  CallExpr* checkDom = new CallExpr("==", actualDom, domExpr->copy());
+
+  VarSymbol* cond = newTemp();
+  retVar->insertBefore(new DefExpr(cond));
+
+  VarSymbol* checkVar = newTemp();
+  retVar->insertBefore(new DefExpr(checkVar));
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, checkVar, checkDom));
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, cond,
+                                    new CallExpr("_cond_test", checkVar)));
+
+  Symbol* msg = new_StringSymbol("domain mismatch on return");
+  BlockStmt* failureMode = new BlockStmt(new CallExpr("halt",
+                                                      msg));
+
+  retVar->insertBefore(new CondStmt(new SymExpr(cond),
+                                    new BlockStmt(new CallExpr(PRIM_NOOP)),
+                                    failureMode));
+}
+
+// Validates the declared element type, if it exists
+static void insertElementTypeCheck(Expr* declaredRet, Expr* actualRet,
+                                   CallExpr* retVar) {
+  VarSymbol* unrefTemp = newTemp();
+  retVar->insertBefore(new DefExpr(unrefTemp));
+  CallExpr* makeArray = new CallExpr(PRIM_MOVE, unrefTemp,
+                                     new CallExpr("chpl__unref",
+                                                  actualRet->copy()));
+  retVar->insertBefore(makeArray);
+
+  VarSymbol* actualEltType = newTemp();
+  actualEltType->addFlag(FLAG_MAYBE_TYPE);
+  actualEltType->addFlag(FLAG_MAYBE_PARAM);
+  retVar->insertBefore(new DefExpr(actualEltType));
+  CallExpr* exprEltType = new CallExpr("eltType", gMethodToken, unrefTemp);
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, actualEltType, exprEltType));
+
+  CallExpr* checkEltType = new CallExpr("==", actualEltType,
+                                        declaredRet->copy());
+
+  VarSymbol* cond = newTemp();
+  cond->addFlag(FLAG_MAYBE_PARAM);
+
+  retVar->insertBefore(new DefExpr(cond));
+
+  VarSymbol* checkVar = newTemp();
+  checkVar->addFlag(FLAG_MAYBE_PARAM);
+  retVar->insertBefore(new DefExpr(checkVar));
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, checkVar, checkEltType));
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, cond,
+                                    new CallExpr("_cond_test", checkVar)));
```

Comment:
Once again I think it'd really improve this code to call a modules routine like `chpl_check_array_eltTypes_match` and have that issue the `compilerError` if necessary.